### DISCUSSION TITLE
Access to configuration variables (updated proposal)

### DISCUSSION THREAD

**Discussion by Olen** [home-assistant/architecture#533] [Entity Models]
This is to pick up where #459 left, and to build on some of the comments there.

#### Summary
What is needed on a high level is a way to add some "settings" to an entity. These settings are entity specific, and can not be retrieved from the "physical" entity itself (or by any other means from an external source).
And we need a way to expose these settings to the rest of HA - not only to the GUI, but also to be used in templates and automations etc.

### Problem description

We have several components where there are configured values, which in reality are more attributes than actual config of the entities.  E.g. they do not influence on how the physical entity actually behaves. Also, you can not read these values from the entity itself, so the only way to set these values today is to set them as part of the configuration of the entity.
- The plant component is the one I have been working with, where each plant will have max/min values for moisture, temperature etc. In addition to references to what sensors are used to measure different kind of data (like the device_tracker config of a `person`)
- The manual alarm, where access to a configured timeout would be benefitial
- Other examples have come up during various discussion

These values can not be pulled from the physical devices or entities that are set up, as - at least for the plants - there is nothing to actually poll.  They are just virtual devices, in the same manner as a "person" object in HA.  So that means the values need to be _configured_ by the user somehow.

The problem is that by using regular configuration parameters, there is no way to expose these values to templates, automations, scripts etc. except by setting them as attributes, and that is not allowed.

### Original suggestion

The suggested solution in the original Issue #459 is to add another entry to each entity called "conf" which allows an integration to publish parts of its config in the same way it publishes attributes. That way they can be used both in the GUI, but also in templates, automations, scripts etc.

One could imagine having access to `{{ conf('plant.12345', 'max_moisture') }}` the same way we have access to `{{ state_attr('plant.12345', 'moisture') }}`

Of course these should **not** be automatically created for all config parameters, as the configuration can include sensitive data, so only parameters that are explicitly added to this entry will be exposed.

### Option 2

Another option, which I have been thinking about after some of the comments in the original issue, but I want to run it through this discussion before starting to work on it.

#### Background
What if we think of the values that we need to expose here as more like for instance the configured device_trackers of a `person`. It is most definitely something that is _configureable_ so it is not just a regular attribute that is read from an entity.  But they are not neccesarily set in the configuration while setting up the entity.  Although in the case of the plant integration, some sane default values would probably be smart to configure when the integration is set up.
It should also be possible to modify them after the integration is set up.

The heart of problem is that there is nowhere in the actual physical device to save these values, as I have not found a good way to neither store data in my plants or read anything useful from them... And the sensors that are used are just "dumb" read only sensors that can not be configured to store the data.  They just broadcast a buch of values, but have no idea if these values are good or bad for the actual plant they are monitoring. 
So if we want to restore the max/min-values after a restart (and we really want to do that), they need to be saved somewhere inside HA.

Some integrations use the "capabilities" in the entity_registry to store similar data (for instance we have max_temp and min_temp of a `climate` entity in the capability of that entity). 

#### New suggestion - expose capabilities

However, as far as I know, you can not access the `capabilities` of a device in a simple way in templates or automations.  Since that is the goal here, my second suggestion would be to start exposing `capabilities` in templates the same way we expose attributes:

That would give us something like `{{ capability_attr('plant.12345', 'max_moisture') }}` along side  `{{ state_attr('plant.12345', 'moisture') }}`

A binary_sensor can then easily be created for e.g. 
`{{ state_attr('plant.12345', 'moisture') > capability_attr('plant.12345', 'max_moisture') }}`

It would also expose the capabilities from the existing integrations,  E.g. `{{ capability_attr('climate.livingroom', 'max_temp') }}` would automatically exist without having to modify other integrations.
At least in some cases, I imagine visibility of these capabilities could be quite useful for other integrations as well, so it would give more flexibility when writing automations and template sensors etc.

The advantage of this suggestion over the initial suggestion is that `capability_attributes` already exists and all that is needed is to add it to the template helpers.  A potential issue could be if any integration used capabilities to store data that should NOT be available outside of the integrations, but as far as I know, this data does not contain anything secret in any of the integrations I have looked at.

### Option 3

A third option would be more of a work around.  This will just add a service call to set the values we need, and still add them as state attributes once they are set.  This means that we still do not save statically configured values as part of the state-object, but it does not change much else in how the data is stored or accessed.  It is definitely the easiest way, but since there must be a reason you do not allow static config in the state object, it feels more of a "loophole" than an actual solution.
Would it be "more permitted" to save these as attributes of an entity if the values are modifiable after the integration has been set up instead of statically configured?
If you think this is a better way, I am open for that as well.

**Comment by frenck** [comment]
IMHO, Configuration can contain sensitive data and should not be accessible this way.
I don't think this should be in our state machine.

**Reply by Olen** [reply]
I am not suggesting to expose all configuration.  I suggest to either expose a very limited part of the config (explicitly defined in the integration) or exposing `capability_attributes`

**Reply by frenck** [reply]
but configuration and capabilities are not the same things?

**Reply by Olen** [reply]
I don't know where the values that are needed to expose would fit best.  They must be _configured_ (e.g. they can not be read directly from an external source).  But they are more of a description of an entitys "capabilities" than actually how the entity is supposed to function.
But in either way, the core here is to be able to read these values in templates and automations. How HA actually stores the values is not really important to me, hence the two different options.

**Reply by slaxor505** [reply]
@frenck can you answer the question please because we are kind of stuck with development of a plant monitor integration. 
Thank you in advance!

**Reply by frenck** [reply]
I answered in the first reply, my opinion hasn't changed for that matter.

**Reply by slaxor505** [reply]
@frenck Would you be able suggest solution for the problem as none of 3 proposed options are acceptable as it appears. As it seems for me that we have no way to move forward with this kind of component. I'd appreciate your input here. Thanks!

**Reply by frenck** [reply]
I'm sorry, I don't have a solution for you.

**Comment by thomasloven** [comment]
Nothing is stopping your integration from transfering the value of select configuration values to attributes, right?
And for user configurability, that could be either separate entities (`number` e.g.) or just handled by the normal customization methods.

I'm probably missing something, but I can't see how that does not solve the described problem.

**Reply by Olen** [reply]
This is the core of the problem.  We first created a PR to do exactly that.  But the PR was rejected because we are not allowed to expose configuration data as attributes in an entity.

https://github.com/home-assistant/core/pull/42420#pullrequestreview-526697754

(And the same for e.g. this which tried to do the same)
https://github.com/home-assistant/core/pull/41872#pullrequestreview-526683427

So this proposal is suggesting different methods to allow fo limited access to configuration data from other integrations without adding new attributes.

**Reply by thomasloven** [reply]
I see. I did not know about that rule.

**Reply by thecode** [reply]
@Olen not directly answering your question, but from my experience sometime you need to think of another approach.
I suggest to try to explain your problem and discuss it with developers, the best place for it is discord. I had few occasions that with help from others I ended up in a better solution than my original plan. Even now I am working on a PR that was ready (and could be) to be merged but from talking to other developers we found out a better solution is already implemented by other integrations and I am going to rewrite that part in the PR.

**Reply by r-t-s** [reply]
Adding mostly static information  as state attributes adds an unjustified burden to the state engine. 
Configuration information todate has been deemed to be scoped as internal/private. 
A cleanly define concept is  needed. This might be something like adding a  "meta_data"  dictionary to entities.
An integration can choose to populate the "meta_data"  from the configuration file or as it discovers it. 
Then you should be able to to add a template method  {{ meta_data(plant.12345', 'max_moisture') }}
Anything added to meta_data is public.

**Comment by frenck** [comment]
This architecture issue is old, stale, or possibly obsolete. Things changed a lot over the years.

For that reason, I'm going to close this discussion for now to clean up a little.

../Frenck
